reparse.go

```diff
@@ -129,3 +140,72 @@ func EncodeReparsePoint(rp *ReparsePoint) []byte {
 	binary.Write(&b, binary.LittleEndian, target16)
 	return b.Bytes()
 }
+
+// ReparseDataBuffer corresponds to REPARSE_DATA_BUFFER struct defined here:
+// https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/ns-ntifs-_reparse_data_buffer.
+// ReparesPoint struct (and corresponding Encode/Decode functions) provided above only
+// handle symlink or mountpoint type of reparse points. To work with other reparse points
+// use this struct.  When calling SetReparsePoint, if it is a symlink or mountpoint
+// reparse point pass in the buffer generated by EncodeReparsePointData. If it is some
+// other kind of reparse point pass in the buffer generated by
+// ReparseDataBuffer.Encode. (same goes for GetReparsePoints)
+type ReparseDataBuffer struct {
+	ReparseTag        uint32
+	ReparseDataLength uint16
+	Reserved          uint16
+	// a generic data buffer
+	DataBuffer []byte
+}
+
+// Encode encodes the current ReparseDataBuffer struct that can be passed
+// to the SetReparsePoint.
+func (r *ReparseDataBuffer) Encode() []byte {
+	var buf bytes.Buffer
+	binary.Write(&buf, binary.LittleEndian, r.ReparseTag)
+	binary.Write(&buf, binary.LittleEndian, r.ReparseDataLength)
+	binary.Write(&buf, binary.LittleEndian, r.Reserved)
+	buf.Write(r.DataBuffer)
+	return buf.Bytes()
+}
+
+// Decode decodes the byte array (returned by GetReparsePoint) into
+// the current ReparseDataBuffer struct.
+func (r *ReparseDataBuffer) Decode(data []byte) error {
+	rdr := bytes.NewReader(data)
+	err := binary.Read(rdr, binary.LittleEndian, r)
+	if err != nil {
+		fmt.Errorf("failed to decode reparse data buffer: %w", err)
+	}
+	return nil
+}
+
+// GetReparsePoint returns a byte array that represents a ReparseDataBuffer if
+// the file at `path` is a valid reparse point.
+func GetReparsePoint(path string) ([]byte, error) {
+	utf16DestPath, err := windows.UTF16FromString(path)
+	if err != nil {
+		return nil, err
+	}
+	// We need to open the file with windows specific permissions so just using
+	// os.Open won't work.
+	fileHandle, err := windows.CreateFile(&utf16DestPath[0], windows.GENERIC_READ, 0, nil, windows.OPEN_EXISTING, (windows.FILE_ATTRIBUTE_NORMAL | windows.FILE_FLAG_OPEN_REPARSE_POINT | windows.FILE_FLAG_BACKUP_SEMANTICS), 0)
+	if err != nil {
+		return nil, fmt.Errorf("open file failed with: %w", err)
+	}
+	defer windows.Close(fileHandle)
+
+	outBuf := make([]byte, _MAXIMUM_REPARSE_DATA_BUFFER_SIZE)
+	var outBufLen uint32
+	err = windows.DeviceIoControl(fileHandle, _FSCTL_GET_REPARSE_POINT, nil, 0, &outBuf[0], _MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &outBufLen, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get reparse point for file %s: %w", path, err)
+	}
+
+	return outBuf[:outBufLen], nil
+}
+
+// SetReparsePoint sets a reparse point with `data` on the file/directory represented by
+// `handle` . `data` should be an encoded ReparseDataBuffer struct.
+func SetReparsePoint(handle windows.Handle, data []byte) error {
```

Comment:
I still think the split here in API between Get taking in a path and Set using a handle is strange. Maybe we expose a method to call CreateFile with the right permissions you need to get/set and have these all take in handles. Or document what flags you need for CreateFile, but it seems like quite a few